File: java/leetcode/src/Task45.java
import org.junit.jupiter.api.Test;

import java.util.Arrays;

public class Task45 {
    //dp
    public int jump(int[] nums) {
        if (nums.length < 2) {
            return 0;
        }

        //init
        int[] dp = new int[nums.length];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            int max = Math.min(nums[i] + i, nums.length - 1);
            for (int j = i + 1; j <= max; j++) {
                dp[j] = Math.min(dp[j], dp[i] + 1);
            }
        }

        return dp[nums.length - 1];
    }

    public int jump2(int[] nums) {
        if (nums.length < 2) {
            return 0;
        }
        int cnt = 1;
        int cur = 0;

        while (cur + nums[cur] < nums.length-1) {
            int pos = cur;
            for (int i = cur + 1; i <= cur + nums[cur]; i++) {
                if(pos+nums[pos] < i+nums[i]){
                    pos = i;
                }
            }
            cur = pos;
            ++cnt;
        }

        return cnt;
    }

    public int jump3(int[] nums) {
        if (nums.length < 2) {
            return 0;
        }
        int end = 0;
        int cnt = 0;
        int maxPos = 0;

        for (int i = 0; i < nums.length-1; i++) {
            maxPos = Math.max(i+nums[i],maxPos);
            if(maxPos >= nums.length-1){
                ++cnt;
                break;
            }
            if(end == i){
                ++cnt;
                end = maxPos;
            }

        }


        return cnt;
    }


    @Test
    void test() {
        int[] nums = new int[]{2, 3, 1, 1, 4};
        System.out.println(jump2(nums));
    }
}

File: py/Task45.py
from typing import List


class Solution:
    # 1 dp
    def jump(self, nums: List[int]) -> int:
        l = len(nums)
        if l < 2:
            return 0
        dp = [999999999 for _ in range(l)]
        dp[0] = 0
        for i in range(l):
            for j in range(i+1, min(i+nums[i]+1, l)):
                dp[j] = min(dp[i] + 1, dp[j])
        return dp[-1]

    # 2 greedy
    def jump2(self, nums: List[int]) -> int:
        l = len(nums)
        if l < 2:
            return 0
        cnt = preMax = maxPos = 0
        for i, num in enumerate(nums):
            maxPos = max(maxPos, i+num)
            if maxPos >= l-1:
                return cnt+1
            if i == preMax:
                preMax = maxPos
                cnt += 1
        return cnt

    # greed 2020-05-04

    def jump3(self, nums: List[int]) -> int:
        l = len(nums)
        if l < 2:
            return 0
        res = 1
        curMax = preMax = -1
        for i, num in enumerate(nums):
            curMax = max(curMax, i + num)
            if curMax >= l - 1:
                return res
            if i >= preMax:
                res += 1
                preMax = curMax  
        return res


nums = [2, 3, 1, 1, 4]

obj = Solution()
print(obj.jump3(nums))
